global.go

```diff
@@ -0,0 +1,328 @@
+// as specified by method.Needs. Global will panic if these are not met. The
+// method can be determined automatically from the supplied problem which is
+// described below.
+//
+// If p.Status is not nil, it is called before every evaluation. If the
+// returned Status is not NotTerminated or the error is not nil, the
+// optimization run is terminated.
+//
+// The third argument contains the settings for the minimization. The
+// DefaultGlobalSettings function can be called for a Settings struct with the
+// default values initialized. If settings == nil, the default settings are used.
+// Global optimization methods typically do not make assumptions about the number
+// and location of local minima. Thus, the only convergence metric used is the
+// function values found at major iterations of the optimization. Bounds on the
+// length of optimization are obeyed, such as the number of allowed function
+// evaluations.
+//
+// The final argument is the optimization method to use. If method == nil, then
+// an appropriate default is chosen based on the properties of the other arguments
+// (dimension, gradient-free or gradient-based, etc.).
+//
+// If method implements Statuser, method.Status is called before every call
+// to method.Iterate. If the returned Status is not NotTerminated or the
+// error is non-nil, the optimization run is terminated.
+//
+// Global returns a Result struct and any error that occurred. See the
+// documentation of Result for more information.
+//
+// Be aware that the default behavior of Global is to find the minimum.
+// For certain functions and optimization methods, this process can take many
+// function evaluations. If you would like to put limits on this, for example
+// maximum runtime or maximum function evaluations, modify the Settings
+// input struct.
+//
+// Something about Global cannot guarantee strict bounds on function evaluations,
+// iterations, etc. in the precense of concurrency.
+func Global(p Problem, dim int, settings *Settings, method GlobalMethod) (*Result, error) {
+	if p.Func == nil {
+		panic("optimize: objective function is undefined")
+	}
+	if dim <= 0 {
+		panic("optimize: impossible problem dimension")
+	}
+	startTime := time.Now()
+	if method == nil {
+		method = &GuessAndCheck{}
+	}
+	if err := p.satisfies(method); err != nil {
+		return nil, err
+	}
+	if p.Status != nil {
+		_, err := p.Status()
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	if settings == nil {
+		settings = DefaultSettingsGlobal()
+	}
+
+	if settings.Recorder != nil {
+		// Initialize Recorder first. If it fails, we avoid the (possibly
+		// time-consuming) evaluation at the starting location.
+		err := settings.Recorder.Init()
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	stats := &Stats{}
+	optLoc := newLocation(dim, method)
+	optLoc.F = math.Inf(1)
+
+	if settings.FunctionConverge != nil {
+		settings.FunctionConverge.Init(optLoc.F)
+	}
```

Comment:
Ok, but then it will panic in `checkConvergenceGlobal`, so perhaps there should be a check for non-nil there.